Heat Capacity Refactor and Gas Caching

Pickle-Coding opened the pull request:
## About The Pull Request
Tracks the heat capacity of gas mixtures into a heat_capacity var. The heat capacity can be found by simply accessing that var instead of using a proc that calculates it by going through every gas EVERY SINGLE TIME. Adds the set_moles() and change_moles() procs, which changes the moles of a gas and automatically adjust the heat capacity.

Also caches individual gases when they are accessed multiple times. Might reduce the amount of list accesses when trying to access said gas. Also looks cleaner.

Drafted. I need to migrate everything that attempts to arbitrarily modify moles with the new procs. There's a lot of those. Also I tracked the heat capacity change in share() in a really crappy way which isn't good performance wise. I should probably look into that to find a better way before merge. There could be an accuracy drift with some behavior that rounds the moles. Maybe I could look into that to reduce the drift.

- [x] Migrate everything that modifies moles arbitrarily to use setter procs.
- [ ] Make share() better.
- [ ] Attempt to reduce accuracy drift due to quantized moles.
## Why It's Good For The Game
It's simply better this way.
## Changelog
:cl:
code: Heat capacity is tracked.
code: Cache individual gases when accessed multiple times.
fix: Fix ethereal lungs not conserving matter and energy properly.
/:cl:

Pickle-Coding committed:
Track heat capacity, cache individual gases.

Tracks the heat capacity of a gas mixture into the heat_capacity var. We access that var instead of calculating the heat capacity with heat_capacity() whenever we want to find the heat capacity.

Atmospherics processing will manually track the heat capacity changes, while things outside of atmospherics should use the gas_mixture/set_moles() and gas_mixture/change_moles() procs if they want to change the moles of a gas.

Also caches the individual gases from the gas list when accessed multiple times. Makes things cleaner and probably reduces how many lists we have to access from.

Migrates some of the external things interacting with atmos to use set_moles() and change_moles(), but there's quite a lot so I'll do the rest in a future commit(s).

Pickle-Coding committed:
Fixes share() not adjusting heat capacity properly.

Fixes share() not adjusting the heat capacity of the mixtures properly. This might not be the best solution performance wise, but at least it's not broken right now. I'll try to come up with a better way later.

Also migrates more random shit to use the setter procs instead of manually modifying moles. There's still lots more to do.

Pickle-Coding commented:
I resolved merge conflicts like 3 minutes ago. What the fuck?

Pickle-Coding committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into i-hope-this-works

Pickle-Coding commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -382,12 +419,11 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
-				heat_capacity_self_to_sharer += gas_heat_capacity
-			else
-				heat_capacity_sharer_to_self -= gas_heat_capacity //subtract here instead of adding the absolute value because we know that delta is negative.
+		var/gas_heat_capacity = delta * gas[GAS_META][META_GAS_SPECIFIC_HEAT]
+		if(delta > 0)
+			heat_capacity_self_to_sharer += gas_heat_capacity
+		else
+			heat_capacity_sharer_to_self -= gas_heat_capacity //subtract here instead of adding the absolute value because we know that delta is negative.
```
I should probably just use a portion of the turfs heat capacity instead of doing this. I'll look into it later.

Pickle-Coding commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -211,28 +229,36 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 	for(var/id in cached_gases)
-		ADD_GAS(id, removed.gases)
+		ADD_GAS(id, removed_gases)
 		removed_gases[id][MOLES] = QUANTIZE(cached_gases[id][MOLES] * ratio)
 		cached_gases[id][MOLES] -= removed_gases[id][MOLES]
 
 	garbage_collect()
 
+	removed.heat_capacity = ratio * heat_capacity
+	heat_capacity *= (1 - ratio)
```
Why do the gas moles get QUANTIZED() here? This could cause the tracked heat capacity to slightly drift from the actual heat capacity.

EliteCreature replied:
I am just guessing here, but I bet this is due to the game-ruining NaN BS that several coders have worked extremely hard to remove.

LemonInTheDark replied:
we quantize anytime we move gas. the quantized amount is the smallest amount of movement we treat as "real".
This might cause a problem with the idea of tracking heat capacity you're right. might need to re-examine either quantize or the heatcap system

Pickle-Coding commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -194,6 +209,9 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 		cached_gases[id][MOLES] -= removed_gases[id][MOLES]
 	garbage_collect()
 
+	removed.heat_capacity = ratio * heat_capacity
+	heat_capacity *= (1 - ratio)
```
Possible tracked heat capacity drift due to the quantize above. Why do we quantize here?

Pickle-Coding commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -195,34 +202,41 @@
+	var/old_heat_capacity = air.heat_capacity
+	plasma_burn_rate = min(plasma_burn_rate, plasma[MOLES], oxygen[MOLES] * INVERSE(oxygen_burn_ratio)) //Ensures matter is conserved properly
+	plasma[MOLES] = QUANTIZE(plasma[MOLES] - plasma_burn_rate)
+	oxygen[MOLES] = QUANTIZE(oxygen[MOLES] - (plasma_burn_rate * oxygen_burn_ratio))
 	if (super_saturation)
 		ASSERT_GAS(/datum/gas/tritium, air)
-		cached_gases[/datum/gas/tritium][MOLES] += plasma_burn_rate
+		var/list/tritium = cached_gases[/datum/gas/tritium]
+		tritium[MOLES] += plasma_burn_rate
+		air.heat_capacity += plasma_burn_rate * (tritium[GAS_META][META_GAS_SPECIFIC_HEAT] - plasma[GAS_META][META_GAS_SPECIFIC_HEAT] - oxygen[GAS_META][META_GAS_SPECIFIC_HEAT] * oxygen_burn_ratio)
```
Possible heat capacity drift, but this one can easily be fixed. Not sure why this is one of the few reactions that quantizes moles though.

Pickle-Coding committed:
Use cached gases in noblium formation.

I left this uncommited for like a couple of days and forgot what I was planning to do with this commit. This one simply uses the cached gas variables in noblium formation for the sanity check.

Pickle-Coding committed:
Immutable mixtures, electrolyzers, more conversion.

The commit title length is so strict.

This adds an initial_heat_capacity for immutable gas mixtures. The heat_capacity gets set to that during garbage collect. For planetary mixtures, the initial_heat_capacity gets set to the heat_capacity during parse_string_immutable().

Electrolyzer reactions. Caches the gas list. Caches individual gases. Change the air.heat_capacity with the calculated heat capacity change.

change_turf.dm's /turf/open/proc/Assimilate_Air(). Simply sets the total.heat_capacity to the heat_cap / turflen. This allows the turfs' new mixtures have the correct heat capacity when copying from the total.

Also sanitises the mole change in /datum/gas_mixture/proc/change_moles(). Don't have to worry about the min checks and stuff all around the codebase, and instead just in that proc.

For the rest of the stuff, it's just more conversions from arbitrary mole modification to using the /datum/gas_mixture/change_moles() proc.

Pickle-Coding committed:
Silly me.

Use the air_mixture var instead of the nonexistant air var for /datum/electrolyzer_reaction/h2o_conversion/react().

Fixes brainfart in /obj/item/flamethrower/proc/default_ignite().

LemonInTheDark reviewed (commented):
I don't have the time to review this rn, but this is very much something I'm interested in, partially because it makes the transition to airgroups easier, but also just for the idea itself.

Will go over when I'm back

Pickle-Coding committed:
Fix ethereal electrolysis conservation.

Fixes ethereal lungs electrolysis not conserving matter and energy properly. Adds conserve_energy proc for set_moles() and change_moles(). Conserves energy if true.

Also converts the rest into using the set_moles() and change_moles() procs. Yay!

Pickle-Coding committed:
Whoops.

Replace the nonexistant heat_capacity var with gas_mixture.heat_capacity.

Pickle-Coding commented:
I'm not sure if I can make share() better. Anyways everything should be converted now so I'll leave it ready for review.

After a few experiments, the difference between reality is roughly around 0.0001% error. It grows to around 0.00075% error after plasma combustion. Probably due to quantize. The error with anything using quantize could get greater at lower moles.

Pickle-Coding commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -66,10 +64,31 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
+	moles = max(-cached_gas[MOLES], moles)
+	cached_gas[MOLES] += moles
+	heat_capacity += moles * cached_gas[GAS_META][META_GAS_SPECIFIC_HEAT]
+	if(conserve_energy)
+		temperature = temperature * old_heat_capacity / heat_capacity
```
```suggestion
	if(conserve_energy && heat_capacity > MINIMUM_HEAT_CAPACITY)
		temperature = max(temperature * old_heat_capacity / heat_capacity, TCMB)
```

Pickle-Coding commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -66,10 +64,31 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
+	var/old_heat_capacity = heat_capacity
+	heat_capacity += cached_gas[GAS_META][META_GAS_SPECIFIC_HEAT] * (moles - cached_gas[MOLES])
+	cached_gas[MOLES] = moles
+	if(conserve_energy)
+		temperature = temperature * old_heat_capacity / heat_capacity
```
```suggestion
	if(conserve_energy && heat_capacity > MINIMUM_HEAT_CAPACITY)
		temperature = max(temperature * old_heat_capacity / heat_capacity, TCMB)
```

Pickle-Coding committed:
Update code/modules/atmospherics/gasmixtures/gas_mixture.dm

Pickle-Coding committed:
Update code/modules/atmospherics/gasmixtures/gas_mixture.dm
